충남대학교 BE 김재혁 위시리스트 2단계

dmdkr55 opened the pull request:
안녕하세요 멘토님. 
위시리스트 2단계 - 회원 로그인 코드 리뷰 요청드립니다.
step2 branch를 생성하지 않고 과제를 진행하여, 커밋 기록들을 따로 step2로 분리해보려고 하였지만 잘 되지않았습니다. 코드리뷰 요청이 늦어져 죄송합니다.

마지막 2개의 커밋이 2단계에 진행한 내용입니다.
[feat: 회원가입 기능 구현]
[feat: 로그인 기능 구현]

dmdkr55 committed:
feat: 기본 코드 준비

dmdkr55 committed:
feat: 유효성 검사

dmdkr55 committed:
feat: 예외 처리

- GlobalExceptionHandler
예외 발생시 응답 메시지 제공

dmdkr55 committed:
feat: 테스트 코드 추가

- ProductControllerTest
유효성 검사 및 예외 처리에 대한 테스트 코드 추가

dmdkr55 committed:
Merge branch 'dmdkr55' into step1

dmdkr55 committed:
feat: Admin 화면에서의 유효성 검사 추가

- ProductAdminController
유효성 검사 및 예외 처리에 대한 코드 추가
-create-form.html, edit-form.html
예외 처리에 따른 Exception 메시지

dmdkr55 committed:
Merge branch 'step1' of https://github.com/dmdkr55/spring-gift-wishlist into step1

dmdkr55 committed:
refactor: 이미지 URL 유효성 검사

- Product.java
이미지 URL에 대한 pattern을 지정

dmdkr55 committed:
refactor: 상품이름 유효성 검사 수정

- Product.java
'카카오'가 포함된 이름은 needsMdApproval 필드를 true로 하여, MD의 승인이 필요하도록 수정.

dmdkr55 committed:
refactor: 변경된 내용에 따라 테스트 코드 수정
          없는 id의 상품 단건 조회시 오류 수정.

- ProductControllerTest.java
수정된 유효성 검사에 따라 테스트 코드 수정
존재하지 않는 상품 조회에 대한 테스트 추가
- ProductRepository.java
존재하지 않는 상품 조회에 대한 에러 처리

dmdkr55 committed:
feat: 회원가입 기능 구현

dmdkr55 committed:
feat: 회원가입 기능 구현

dmdkr55 committed:
feat: 로그인 기능 구현

dmdkr55 committed:
feat: 로그인 기능 구현

dmdkr55 committed:
Merge branch 'dmdkr55' into step2

sumniy commented on README.md:
```diff
@@ -15,4 +15,31 @@
     * 그 외 특수 문자 사용 불가
 * "카카오"가 포함된 문구는 담당 MD와 협의한 경우에만 사용할 수 있다.
+---
+## 2단계 - 회원 로그인
+
+### 요구사항
+사용자가 회원 가입, 로그인, 추후 회원별 기능을 이용할 수 있도록 구현한다.
+* 회원은 이메일과 비밀번호를 입력하여 가입한다.
+* 토큰을 받으려면 이메일과 비밀번호를 보내야 하며, 가입한 이메일과 비밀번호가 일치하면 토큰이 발급된다.
+* 토큰을 생성하는 방법에는 여러 가지가 있다. 방법 중 하나를 선택한다.
+* (선택) 회원을 조회, 추가, 수정, 삭제할 수 있는 관리자 화면을 구현한다.
+* Authorization 헤더가 유효하지 않거나 토큰이 유효하지 않은 경우 401 Unauthorized를 반환한다.
+* 잘못된 로그인, 비밀번호 찾기, 비밀번호 변경 요청은 403 Forbidden을 반환한다.
+
+1. 회원가입 기능
+   * 회원 가입 API(/api/members/register) 구현 
+   * 이메일과 비밀번호로 회원 가입 요청을 받을 수 있다. 
+   * 회원 가입 시 중복 이메일은 허용하지 않는다. 
+   * 회원 가입 시 JWT 토큰을 발급하여 응답한다.
+
+2. 로그인 기능
+   * 로그인 API(/api/members/login) 구현
+   * 이메일과 비밀번호가 일치하면 JWT 토큰을 발급한다. 
+   * 존재하지 않는 이메일 또는 비밀번호가 틀리면 403 Forbidden을 반환한다.
+
+3. (선택) 관리자 기능 
+   * 관리자 전용 회원 관리 기능 (추가/조회/수정/삭제 API)
+   * 관리자는 모든 회원을 조회할 수 있다.
+   * 일반 회원은 해당 기능에 접근할 수 없다.
```
깔끔하게 정리 잘 하셨네요 적혀 있는 대로 확인하니 편했습니다 👍

dmdkr55 committed:
feat: 관리자 페이지에서 멤버 추가 기능

sumniy commented on src/main/java/gift/dto/RegisterRequest.java:
```diff
@@ -0,0 +1,5 @@
+package gift.dto;
+
+public record RegisterRequest(String email, String password) {
```
이메일 형식이나 비밀번호 빈 문자열에 대한 유효성 검사가 있으면 좋을 것 같네요!

sumniy commented on src/main/java/gift/exception/GlobalExceptionHandler.java:
```diff
@@ -34,4 +35,13 @@ protected ResponseEntity<Object> handleMethodArgumentNotValid(
         return new ResponseEntity<>(body, HttpStatus.BAD_REQUEST);
     }
+  
+    @ExceptionHandler(IllegalArgumentException.class)
```
`IllegalArgumentException`의 경우에 일반적으로 여러 상황에서 쓰일 수 있는 잘못된 파라미터에 대한 Exception인데 전부 CONFLICT로 반환되네요. 
중복 이메일 가입에 대한 처리가 다른 Exception 처리에 영향을 주지 않도록 하려면 어떻게 하면 좋을까요? 고민해서 구현해 보시면 좋을 것 같아요 😊

sumniy commented on src/main/java/gift/model/Member.java:
```diff
@@ -0,0 +1,41 @@
+
+    private Long id;
+
+    @NotBlank(message = "이메일은 필수 입력 값입니다.")
+    private String email;
+
+    @NotBlank(message = "비밀번호는 필수 입력 값입니다.")
+    private String password;
```
여기 적용된 @NotBlank 는 유효성 검사에 적용되지 않고 있네요. 실제로 빈 문자열로 가입이 되구요. 
validation 관련 어노테이션이 어떻게 동작하는지 찾아보시면 학습에 도움 되실 거예요 😊 
https://mangkyu.tistory.com/174

sumniy commented on src/main/java/gift/util/PasswordUtil.java:
```diff
@@ -0,0 +1,19 @@
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import org.springframework.security.crypto.password.PasswordEncoder;
+
+@Configuration
```
static 메서드만 제공하는 Util 클래스에선 @Configuration 선언이 불필요해 보이는데요. 
@Configuration 가 붙었을 때 어떤 일이 일어나는지 왜 사용하는지 찾아보시면 좋을 것 같습니다 ㅎㅎ

sumniy commented on src/main/java/gift/service/MemberService.java:
```diff
@@ -0,0 +1,62 @@
+            throw new IllegalArgumentException("이미 가입된 이메일입니다: " + request.email());
+        }
+
+        String encryptedPassword = PasswordUtil.encode(request.password());
```
패스워드 암호화 👍

sumniy commented on src/main/java/gift/service/MemberService.java:
```diff
@@ -0,0 +1,62 @@
+        Member member = new Member(request.email(), encryptedPassword);
+        Member savedMember = memberRepository.save(member);
+
+        String secretKey = "Yn2kjibddFAWtnPJ2AFlL8WXmohJMCvigQggaEypa5E=";
```
secretKey는 클래스에도 선언되어 있는데 제거를 깜빡하신 것 같네요! 요기는 지워주세요~

sumniy commented on src/main/java/gift/service/MemberService.java:
```diff
@@ -0,0 +1,62 @@
+
+        String secretKey = "Yn2kjibddFAWtnPJ2AFlL8WXmohJMCvigQggaEypa5E=";
+
+        String accessToken = Jwts.builder()
+            .setSubject(savedMember.getId().toString())
+            .claim("email", savedMember.getEmail())
+            .signWith(Keys.hmacShaKeyFor(secretKey.getBytes()))
+            .compact();
```
토큰 생성 코드가 중복되는데 메서드로 분리하면 좋을 것 같습니다~

sumniy reviewed (changes requested):
재혁님 2단계 구현하느라 고생 많으셨습니다!
전반적으로 요구사항대로 잘 구현해 주셨네요 👍

다만 인증 토큰이 쓰이는 곳이 없어서 조금 수정 요청 드릴게요.
가입이나 로그인 성공 후 받은 토큰이 있어야 admin 페이지를 볼 수 있도록 만들어 주세요!
토큰을 실제로 사용하게 되면서 코드에서 고려할 사항이 많아지니 학습에 도움이 되실 거예요. 화이팅입니다!

dmdkr55 committed:
refactor: 토큰 생성 메소드 분리

- MemberService.java
secretKey 중복 제거.
토큰 생성 중복 코드 메소드로 분리

dmdkr55 committed:
refactor: Exception Handler 추가

- GlobalExceptionHandler.java
회원가입, 로그인 시 발생가능한 예외에 대한 핸들러들을 분리

dmdkr55 committed:
feat: Admin API에 JWT 인증 필터 적용

- JwtAuthenticationFilter.java, FilterConfig.java
Admin API 요청에 대해 JWT 인증 로직 추가

dmdkr55 commented:
코드 수정하여 다시 코드 리뷰 요청드립니다.!

- filter를 적용하여 admin api 요청에 대해 JWT 인증하도록 수정하였습니다.
"가입이나 로그인 성공 후 받은 토큰이 있어야 admin 페이지를 볼 수 있도록"이라고 수정 요청을 해주셨습니다.!
URL을 통해 GET 요청하여 관리자 화면(list.html)을 볼 수 있도록 구현하고자 하였으나, header에 발급된 토큰을 담아서 요청하는 방법을 찾지 못하였습니다.  따라서 관리자 화면 대신 Postman을 통해 admin API 요청 시 헤더에 토큰을 담아 동작을 테스트하였습니다.
로그인 UI를 만들어 로그인을 통해 토큰을 발급 -> 발급받은 토큰을 local storage에 저장 -> 다른 페이지로 이동 요청 시 local storage에 저장된 Token을 헤더에 추가하여 서버에 요청 -> 서버에서 토큰 인증 후 페이지 로딩
토큰 인증을 통해 admin 페이지를 볼 수 있도록 이러한 방식을 찾아보았지만, 이 방식은 Server-side-rendering에서는 힘든 방법이라고 생각이 되었습니다.
혹시 다른 방법이 있는지 궁금합니다..!!
- IllegalArgumentException은 여러 상황에서 쓰일 수 있는 Exception이라고 말씀 주셨습니다.
이미 가입된 이메일로 회원가입, 로그인 시 비밀번호가 일치하지 않는 등 이런 특별한 상황에 대한 Exception 클래스를 생성하여 문제 발생 시 문제 상황을 직관적으로 확인할 수 있도록 수정하였습니다.

감사합니다!

sumniy commented on src/main/java/gift/config/FilterConfig.java:
```diff
@@ -0,0 +1,21 @@
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+public class FilterConfig {
```
Filter를 활용해서 request에 대한 검증을 한 번에 처리하도록 잘 구현하셨습니다 👍

sumniy commented on src/main/java/gift/dto/LoginRequest.java:
```diff
@@ -0,0 +1,33 @@
+public class LoginRequest {
+
+    @NotBlank(message = "이메일는 필수 입력 값입니다.")
+    @Pattern(regexp = "^[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,}$", message = "유효한 이메일 형식이 아닙니다.")
```
이메일 유효성 검사는 정의된 어노테이션을 활용할 수 있어요! 

```suggestion
    @Email(message = "유효한 이메일 형식이 아닙니다.")
```

sumniy commented on src/main/java/gift/exception/GlobalExceptionHandler.java:
```diff
@@ -34,4 +35,32 @@ protected ResponseEntity<Object> handleMethodArgumentNotValid(
+        return createErrorResponse(HttpStatus.BAD_REQUEST, e.getMessage());
+    }
+
+    @ExceptionHandler(EmailAlreadyExistsException.class)
+    public ResponseEntity<?> handleEmailExists(EmailAlreadyExistsException e) {
+        return createErrorResponse(HttpStatus.CONFLICT, e.getMessage());
+    }
+
+    @ExceptionHandler(MemberNotFoundException.class)
+    public ResponseEntity<?> handleMemberNotFound(MemberNotFoundException e) {
+        return createErrorResponse(HttpStatus.NOT_FOUND, e.getMessage());
+    }
+
+    @ExceptionHandler(InvalidPasswordException.class)
+    public ResponseEntity<?> handleInvalidPassword(InvalidPasswordException e) {
+        return createErrorResponse(HttpStatus.BAD_REQUEST, e.getMessage());
+    }
```
에러 상황에 따라서 커스텀 Exception 클래스를 정의해서 각 에러에 맞는 적절한 응답을 하도록 잘 만드셨네요! 💯
현재 시스템에 맞춰서 잘 구현하셨고, 다음 단계로는 확장성에 대한 고민을 해보시면 좋을 것 같아요. 

지금과 같은 방식에서는 에러 상황이 수십, 수백개로 예상될 때 Exception 클래스를 모두 정의하고 handler를 정의해야 하는데, 최소한의 변경으로 예외 처리를 추가할 수 있는 방법을 찾으면 좋겠죠? 

이번 미션에서는 생각만 해보시고 구현은 하지 않아도 좋아요~

sumniy commented on src/main/java/gift/filter/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,53 @@
+
+        String uri = request.getRequestURI();
+
+        // admin 경로만 토큰 검사
```
경로 지정으로 원하는 path에만 filter를 적용하도록 하셨네요 👍

sumniy reviewed (approved):
쉽지 않은 수정 작업이었을텐데 잘 구현해 주셨습니다 😊 고생하셨어요 ㅎㅎ

인증에 대해서는 서버 사이드 렌더링과는 사실 무관하게 생각하신 방법도 가능하고, 쿠키를 이용하는 방법도 있고 몇 가지 방법이 가능하겠지만 클라이언트 사이드의 이해도와 구현 능력이 필요할 거예요.
 
그러니 지금은 공부할 내용으로 남겨두고 Member 관련 페이지와 컨트롤러에서는 인증을 제외하고, 
`/api/products`상품 관련 API에서만 헤더 방식으로 인증하도록 변경하는 방법을 추천드립니다! 

추가로 JWT를 다루는 코드가 여기저기 있어서 secret 키나 토큰 관련 로직이 노출되고, 중복되고 있어요. JWT 로직을 다루는 별도 클래스를 만들어서 활용하도록 수정 부탁드릴게요!

이번 단계는 머지할게요~ 피드백은 다음 단계에 같이 적용해 주세요! 💪
